### PR TITLE
 Rename ismatch to occursin; fix 0.7 depwarns 

### PR DESCRIPTION
**Pull request by jmkuhn**
`ismatch` has been renamed to `occursin` in Julia Base.
https://github.com/JuliaLang/julia/pull/24673
https://github.com/JuliaLang/julia/pull/26283

To accommodate this in Glob, for Julia 0.6 allow
either `ismatch` or `occursin` to be used.
```
julia> using Glob

julia> using Compat

julia> ismatch(fn"AB*AB*AB", "ABXABXAB")
true

julia> occursin(fn"AB*AB*AB", "ABXABXAB")
true
```

With the current 0.7 nightly, `ismatch` produces a deprecation warning.
```
julia> using Glob

julia> ismatch(fn"AB*AB*AB", "ABXABXAB")
┌ Warning: `ismatch(fn::FilenameMatch, s::AbstractString)` is deprecated, use `occursin(fn, s)` instead.
│   caller = top-level scope
└ @ Core :0
true

julia> occursin(fn"AB*AB*AB", "ABXABXAB")
true
```

I'm not sure my updates to the iterator are the best way to do this.  If you have any suggestions, let me know.

### CODE REVIEW

File: src/Glob.jl [hunk 1]
```diff
@@ -90,16 +99,16 @@ function ismatch(fn::FilenameMatch, s::AbstractString)
...
+                        mc, mi = iterate(pattern, mi)
                     end
                     match = ((c == mc) || (caseless && uppercase(c)==uppercase(mc)))
                 end
             end
         end
         if !match # try to backtrack and add another character to the last *
             star == 0 && return false
-            c, i = next(s, starmatch)
+            c, i = iterate(s, starmatch)
```
Comment (vtjnash) [inline src/Glob.jl]:
```julia
    c, i = something(iterate(s, starmatch)) # starmatch is strictly <= i, so it is known that it must be a valid index
```

File: src/Glob.jl [hunk 2]
```diff
@@ -108,65 +117,78 @@ function ismatch(fn::FilenameMatch, s::AbstractString)
...
+
+@static if VERSION < v"0.7.0-DEV.4637"
+    Base.ismatch(fn::FilenameMatch, s::AbstractString) = occursin(fn, s)
+else
+    @deprecate ismatch(fn::FilenameMatch, s::AbstractString) occursin(fn, s)
+end
+
+filter!(fn::FilenameMatch, v) = filter!(x->occursin(fn,x), v)
+filter(fn::FilenameMatch, v)  = filter(x->occursin(fn,x), v)
+filter!(fn::FilenameMatch, d::Dict) = filter!((k,v)->occursin(fn,k),d)
```
Comment (vtjnash) [inline src/Glob.jl]: The `filter!` now calls the function with one argument, so this should be:
```julia
filter!(fn::FilenameMatch, d::Dict) = filter!(((k, v),) -> occursin(fn, k), d)
```

File: src/Glob.jl [hunk 3]
```diff
@@ -179,13 +201,13 @@ function _match_bracket(pat::AbstractString, mc::Char, i, cl::Char, cu::Char) #
             #match = (pat[j:k0] == s[ci:ci+(k0-j)])
             #return (mc, k3, true, match)
         end
-        mc, j = next(pat, j)
+        mc, j = iterate(pat, j)
```
Comment (vtjnash) [inline src/Glob.jl]: wrap in `something()` (like above)

File: src/Glob.jl [hunk 4]
```diff
@@ -179,13 +201,13 @@ function _match_bracket(pat::AbstractString, mc::Char, i, cl::Char, cu::Char) #
...
         return (mc, k3, false, true)
     else #if mc2 == '='
         if j != k0
             error(string("only single characters are currently supported as character equivalents, got [=", SubString(pat, j, k0), "=]"))
         end
-        mc, j = next(pat, j)
+        mc, j = iterate(pat, j)
```
Comment (vtjnash) [inline src/Glob.jl]: wrap in `something()` (like above)

File: src/Glob.jl [hunk 5]
```diff
@@ -200,19 +222,19 @@ function _match(pat::AbstractString, i0, c::Char, caseless::Bool, extended::Bool
         cl = cu = c
     end
     i = i0
-    if done(pat, i)
+    if i > ncodeunits(pat)
         return (i0, false, c=='[')
     end
-    mc, j = next(pat, i)
+    mc, j = iterate(pat, i)
```
Comment (vtjnash) [inline src/Glob.jl]: like you did above, use:
```julia
next = iterate(pat, i)
if next === nothing
    return (i0, false, c=='[')
end
mc, j = next
```

File: src/Glob.jl [hunk 6]
```diff
@@ -200,19 +222,19 @@ function _match(pat::AbstractString, i0, c::Char, caseless::Bool, extended::Bool
...
     if mc == '!'
         negate = true
         i = j
     end
     match = false
     notfirst = false
-    while !done(pat,i)
-        mc, i = next(pat, i)
+    while i <= ncodeunits(pat)
+        mc, i = iterate(pat, i)
```
Comment (vtjnash) [inline src/Glob.jl]: similar to above, use:
```julia
while true
    next = iterate(pat, i)
    next === nothing && break
    mc, i = next
...
end
```

File: src/Glob.jl [hunk 7]
```diff
@@ -226,20 +248,20 @@ function _match(pat::AbstractString, i0, c::Char, caseless::Bool, extended::Bool
                 return (i0, false, c=='[')
             end
         elseif extended & (mc == '\\')
-            if done(pat, i)
+            if i > ncodeunits(pat)
```
Comment (vtjnash) [inline src/Glob.jl]: similarly for all of these lines, use the assignment to the temporary `next` variable pattern for these that follow here

**Comment by jmkuhn** [comment]
Thanks for the thorough review.  The iterator code looks much better now.

**Pull request merged**